rotkehlchen/db/dbhandler.py

```diff
@@ -231,26 +236,67 @@ def __init__(
         self.user_data_dir = user_data_dir
         self.sql_vm_instructions_cb = sql_vm_instructions_cb
         self.sqlcipher_version = detect_sqlcipher_version()
-        self.last_write_ts: Optional[Timestamp] = None
         self.setting_to_default_type = {
             'version': (int, ROTKEHLCHEN_DB_VERSION),
             'last_write_ts': (int, Timestamp(0)),
             'last_data_upload_ts': (int, Timestamp(0)),
             'premium_should_sync': (str_to_bool, DEFAULT_PREMIUM_SHOULD_SYNC),
             'main_currency': (lambda x: Asset(x).resolve(), A_USD.resolve_to_fiat_asset()),
+            'ongoing_upgrade_from_version': (int, None),
         }
         self.conn: DBConnection = None  # type: ignore
         self.conn_transient: DBConnection = None  # type: ignore
         # Lock to make sure that 2 callers of get_or_create_evm_token do not go in at the same time
         self.get_or_create_evm_token_lock = Semaphore()
         self._connect(password)
+        self._check_abandoned_upgrades(password)
         self._run_actions_after_first_connection(password)
         with self.user_write() as cursor:
             if initial_settings is not None:
                 self.set_settings(cursor, initial_settings)
             self.update_owned_assets_in_globaldb(cursor)
             self.add_globaldb_assetids(cursor)
 
+    def _check_abandoned_upgrades(self, password: str) -> None:
+        """
+        Checks the database whether there are any not finished upgrades and automatically uses a
+        backup if there are any. If no backup found, throws an error to the user
+        """
+        with self.conn.read_ctx() as cursor:
+            try:
+                ongoing_upgrade_from_version = self.get_setting(
+                    cursor=cursor,
+                    name='ongoing_upgrade_from_version',
+                )
+            except sqlcipher.OperationalError:  # pylint: disable=no-member
+                return  # fresh database. Nothing to upgrade.
+        if ongoing_upgrade_from_version is None:
+            return  # We are all good
+
+        # Otherwise replace the db with a backup and relogin
+        self.logout()
+        backup_postfix = f'rotkehlchen_db_v{ongoing_upgrade_from_version}.backup'
+        found_backups = list(filter(
+            lambda x: x[-len(backup_postfix):] == backup_postfix,
+            os.listdir(self.user_data_dir),
+        ))
+        if len(found_backups) == 0:
+            raise DBUpgradeError(
+                'Your encrypted database is in a half-upgraded state and there was no backup '
+                'found. Please open an issue on our github or contact us in our discord server.',
+            )
+
+        backup_to_use = sorted(found_backups)[-1]  # Use latest backup
+        shutil.copyfile(
+            os.path.join(self.user_data_dir, backup_to_use),
+            os.path.join(self.user_data_dir, 'rotkehlchen.db'),
+        )
+        self.msg_aggregator.add_warning(
```

Comment:
This functionality could also be in another PR. We should talk about this and probably have the user agree to this from the UI. This is a bit more complicated than this.